// memoize@11.0.0/distribution/index.js
import mimicFunction from 'mimic-function';
const maxTimeoutValue = 2_147_483_647;
const cacheStore = new WeakMap();
const cacheTimerStore = new WeakMap();
const cacheKeyStore = new WeakMap();
function getValidCacheItem(cache, key) {
    const item = cache.get(key);
    if (!item) {
        return undefined;
    }
    if (item.maxAge <= Date.now()) {
        cache.delete(key);
        return undefined;
    }
    return item;
}
function validateMaxAge(value, source) {
    if (value === Number.POSITIVE_INFINITY) {
        return;
    }
    if (!Number.isFinite(value)) {
        if (source === '`maxAge` option') {
            throw new TypeError('The `maxAge` option must be a finite number, `0`, or `Infinity`.');
        }
        throw new TypeError('The `maxAge` function must return a finite number, `0`, or `Infinity`.');
    }
    if (value > maxTimeoutValue) {
        if (source === '`maxAge` option') {
            throw new TypeError(`The \`maxAge\` option cannot exceed ${maxTimeoutValue}.`);
        }
        throw new TypeError(`The \`maxAge\` function result cannot exceed ${maxTimeoutValue}.`);
    }
}
function getPropertyDescriptor(object, property) {
    let currentObject = object;
    while (currentObject) {
        const descriptor = Object.getOwnPropertyDescriptor(currentObject, property);
        if (descriptor) {
            return descriptor;
        }
        const prototype = Object.getPrototypeOf(currentObject);
        currentObject = prototype === null ? undefined : prototype;
    }
    return undefined;
}
/**
[Memoize](https://en.wikipedia.org/wiki/Memoization) functions - An optimization used to speed up consecutive function calls by caching the result of calls with identical input.

@param function_ - The function to be memoized.

@example
```
import memoize from 'memoize';

let index = 0;
const counter = () => ++index;
const memoized = memoize(counter);

memoized('foo');
//=> 1

// Cached as it's the same argument
memoized('foo');
//=> 1

// Not cached anymore as the arguments changed
memoized('bar');
//=> 2

memoized('bar');
//=> 2
```
*/
export default function memoize(function_, { cacheKey, cache = new Map(), maxAge, } = {}) {
    if (typeof maxAge === 'number') {
        validateMaxAge(maxAge, '`maxAge` option');
        if (maxAge <= 0) {
            return function_;
        }
    }
    const memoized = function (...arguments_) {
        const key = cacheKey ? cacheKey(arguments_) : arguments_[0];
        const cacheItem = getValidCacheItem(cache, key);
        if (cacheItem) {
            return cacheItem.data;
        }
        const result = function_.apply(this, arguments_);
        const computedMaxAge = typeof maxAge === 'function' ? maxAge(...arguments_) : maxAge;
        if (computedMaxAge !== undefined && computedMaxAge !== Number.POSITIVE_INFINITY) {
            validateMaxAge(computedMaxAge, '`maxAge` function result');
            if (computedMaxAge <= 0) {
                return result; // Do not cache
            }
        }
        cache.set(key, {
            data: result,
            maxAge: (computedMaxAge === undefined || computedMaxAge === Number.POSITIVE_INFINITY)
                ? Number.POSITIVE_INFINITY
                : Date.now() + computedMaxAge,
        });
        if (computedMaxAge !== undefined && computedMaxAge !== Number.POSITIVE_INFINITY) {
            const timer = setTimeout(() => {
                cache.delete(key);
                cacheTimerStore.get(memoized)?.delete(timer);
            }, computedMaxAge);
            timer.unref?.();
            const timers = cacheTimerStore.get(memoized) ?? new Set();
            timers.add(timer);
            cacheTimerStore.set(memoized, timers);
        }
        return result;
    };
    mimicFunction(memoized, function_, {
        ignoreNonConfigurable: true,
    });
    cacheStore.set(memoized, cache);
    cacheKeyStore.set(memoized, (cacheKey ?? ((arguments_) => arguments_[0])));
    return memoized;
}
/**
@returns A [decorator](https://github.com/tc39/proposal-decorators) to memoize class methods, static class methods, or getters.

@example
```
import {memoizeDecorator} from 'memoize';

class Example {
    index = 0

    @memoizeDecorator()
    counter() {
        return ++this.index;
    }
}

class ExampleWithOptions {
    index = 0

    @memoizeDecorator({maxAge: 1000})
    counter() {
        return ++this.index;
    }
}

class ExampleWithGetter {
    @memoizeDecorator()
    get value() {
        return expensiveComputation();
    }
}
```
*/
export function memoizeDecorator(options = {}) {
    function decorator(target, context) {
        if (context.kind === 'method') {
            if (context.private) {
                throw new TypeError('The `memoizeDecorator` method decorator does not support private methods.');
            }
            const memoizedMethods = new WeakMap();
            const getMemoizedMethod = (receiver) => {
                let memoizedMethod = memoizedMethods.get(receiver);
                if (!memoizedMethod) {
                    memoizedMethod = memoize(target.bind(receiver), options);
                    memoizedMethods.set(receiver, memoizedMethod);
                }
                return memoizedMethod;
            };
            const installMethod = (receiver, method) => {
                Object.defineProperty(receiver, context.name, {
                    configurable: true,
                    enumerable: false,
                    value: method,
                    writable: true,
                });
                return method;
            };
            const installStaticMethodGetter = (receiver) => {
                Object.defineProperty(receiver, context.name, {
                    configurable: true,
                    enumerable: false,
                    get() {
                        const currentReceiver = this;
                        if (currentReceiver !== receiver) {
                            const currentPrototype = Object.getPrototypeOf(currentReceiver);
                            if (currentPrototype === receiver
                                && !Object.hasOwn(currentReceiver, context.name)) {
                                installStaticMethodGetter(currentReceiver);
                            }
                        }
                        return getMemoizedMethod(currentReceiver);
                    },
                });
            };
            const decoratedMethod = function (...arguments_) {
                return getMemoizedMethod(this)(...arguments_);
            };
            mimicFunction(decoratedMethod, target, {
                ignoreNonConfigurable: true,
            });
            if (context.static) {
                context.addInitializer(function () {
                    installStaticMethodGetter(this);
                });
            }
            else {
                context.addInitializer(function () {
                    const receiver = this;
                    // Checking the descriptor avoids invoking an overriding accessor while `super()` is still running.
                    const descriptor = getPropertyDescriptor(Object.getPrototypeOf(receiver), context.name);
                    if (descriptor?.value !== decoratedMethod) {
                        return;
                    }
                    installMethod(receiver, getMemoizedMethod(receiver));
                });
            }
            return decoratedMethod;
        }
        const memoizedGetters = new WeakMap();
        return function () {
            let memoizedGetter = memoizedGetters.get(this);
            if (!memoizedGetter) {
                memoizedGetter = memoize(() => target.call(this), options);
                memoizedGetters.set(this, memoizedGetter);
            }
            return memoizedGetter();
        };
    }
    return decorator;
}
/**
Clear all cached data of a memoized function.

@param function_ - The memoized function.
*/
export function memoizeClear(function_) {
    const cache = cacheStore.get(function_);
    if (!cache) {
        throw new TypeError('Can\'t clear a function that was not memoized!');
    }
    if (typeof cache.clear !== 'function') {
        throw new TypeError('The cache Map can\'t be cleared!');
    }
    cache.clear();
    for (const timer of cacheTimerStore.get(function_) ?? []) {
        clearTimeout(timer);
    }
    cacheTimerStore.delete(function_);
}
/**
Check if a specific set of arguments is cached for a memoized function.

@param function_ - The memoized function.
@param arguments_ - The arguments to check.
@returns `true` if the arguments are cached and not expired, `false` otherwise.

Uses the same argument processing as the memoized function, including any custom `cacheKey` function.

@example
```
import memoize, {memoizeIsCached} from 'memoize';

const expensive = memoize((a, b) => a + b, {cacheKey: JSON.stringify});
expensive(1, 2);

memoizeIsCached(expensive, 1, 2);
//=> true

memoizeIsCached(expensive, 3, 4);
//=> false
```
*/
export function memoizeIsCached(function_, ...arguments_) {
    const cacheKey = cacheKeyStore.get(function_);
    if (!cacheKey) {
        return false;
    }
    const cache = cacheStore.get(function_);
    const key = cacheKey(arguments_);
    const item = getValidCacheItem(cache, key);
    return item !== undefined;
}
